quantum/impl/quantum_read_write_mutex_impl.h

```diff
@@ -121,177 +144,203 @@ int ReadWriteMutex::numReaders() const
-                                     ReadWriteMutex::AdoptLock) :
+ReadWriteMutex::Guard::Guard(ICoroSync::Ptr sync,
+                             ReadWriteMutex& lock,
+                             LockTraits::AcquireWrite) :
     _mutex(&lock),
-    _ownsLock(true)
+    _ownsLock(true),
+    _isUpgraded(true)
 {
+    _mutex->lockWrite(sync);
 }
 
 inline
-ReadWriteMutex::ReadGuard::~ReadGuard()
+ReadWriteMutex::Guard::Guard(ReadWriteMutex& lock,
+                             LockTraits::AcquireRead,
+                             LockTraits::TryToLock) :
+    _mutex(&lock),
+    _ownsLock(_mutex->tryLockRead()),
+    _isUpgraded(false)
 {
-    unlock();
 }
 
 inline
-void ReadWriteMutex::ReadGuard::lock()
+ReadWriteMutex::Guard::Guard(ReadWriteMutex& lock,
+                             LockTraits::AcquireWrite,
+                             LockTraits::TryToLock) :
+    _mutex(&lock),
+    _ownsLock(_mutex->tryLockWrite()),
+    _isUpgraded(_ownsLock)
 {
-    lock(nullptr);
 }
 
 inline
-void ReadWriteMutex::ReadGuard::lock(ICoroSync::Ptr sync)
+ReadWriteMutex::Guard::Guard(ReadWriteMutex& lock,
+                             LockTraits::AdoptLock) :
+    _mutex(&lock),
+    _ownsLock(lock.isLocked()),
+    _isUpgraded(lock.isWriteLocked())
 {
-    if (_ownsLock) return;
-    assert(_mutex);
-    _mutex->lockRead(sync);
-    _ownsLock = true;
 }
 
 inline
-bool ReadWriteMutex::ReadGuard::tryLock()
+ReadWriteMutex::Guard::~Guard()
 {
-    
-    if (_ownsLock) return true;
-    assert(_mutex);
-    _ownsLock = _mutex->tryLockRead();
-    return _ownsLock;
+    if (ownsLock()) {
+        unlock();
+    }
 }
 
 inline
-void ReadWriteMutex::ReadGuard::unlock()
+void ReadWriteMutex::Guard::lockRead()
 {
-    if (!_ownsLock) return;
-    _mutex->unlockRead();
-    _ownsLock = false;
+    //Application must use the other lock() overload if we are running inside a coroutine
+    assert(!local::context());
+    lockRead(nullptr);
 }
 
 inline
-void ReadWriteMutex::ReadGuard::release()
+void ReadWriteMutex::Guard::lockWrite()
 {
-    _ownsLock = false;
-    _mutex = nullptr;
+    //Application must use the other lock() overload if we are running inside a coroutine
+    assert(!local::context());
+    lockWrite(nullptr);
 }
 
 inline
-bool ReadWriteMutex::ReadGuard::ownsLock() const
+void ReadWriteMutex::Guard::lockRead(ICoroSync::Ptr sync)
 {
-    return _ownsLock;
+    assert(_mutex && !ownsLock());
+    _mutex->lockRead(sync);
+    _ownsLock = true;
+    _isUpgraded = false;
 }
 
-//==============================================================================================
-//                                class WriteGuard
-//==============================================================================================
 inline
-ReadWriteMutex::WriteGuard::WriteGuard(ReadWriteMutex& lock) :
-    ReadWriteMutex::WriteGuard::WriteGuard(nullptr, lock)
+void ReadWriteMutex::Guard::lockWrite(ICoroSync::Ptr sync)
 {
+    assert(_mutex && !ownsLock());
```

Comment:
should assert !ownsWriteLock() as well.

Reply:
ownsLock includes ownsWriteLock.

Reply:
my mistake, this (and others) looks good.